File: js/Neetcode_150/4_Binary_Search/25_search_in_rsa.js
//Search in Rotated Sorted Array: given a RSA & an int 'target', find it and return its index in arr; if target not present in arr, return -1

var search = function(nums, target) {
    let l = 0, r = nums.length-1
    while(l<=r){
        let m = Math.floor(l + (r-l)/2)
        if(nums[m] == target) return m
        //if we're in left part
        if(nums[l] <= nums[m]){
            //if target is also in left part (b/w l & m)
            if(target >= nums[l] && target <= nums[m]) r = m-1
            else l = m+1            
        }
        //if we're in right part
        else{
            //if target is also in right part (b/w m & r)
            if(target >= nums[m] && target <= nums[r]) l = m+1
            else r = m-1
        }
    }
    return -1
};

File: c++/Neetcode_150/4_Binary_Search/25_search_in_rsa.cpp
//Search in Rotated Sorted Array: given a RSA & an int 'target', find it and return its index in arr; if target not present in arr, return -1

int search(vector<int>& nums, int target) {
        int n = nums.size();
        int s = 0, e = n-1;
        while(s<=e){
            int mid = s + ((e-s)/2);
            if(nums[mid] == target) return mid;
            //If we're in left part [s->mid] 
            if(nums[s] <= nums[mid]){
                //if element in this sorted part
                if(nums[s] <= target && target <= nums[mid]){
                    e = mid-1;
                }
                //element not in this sorted part 
                else{
                    s = mid+1;
                }
            }
            //If we're in right part [mid->e]
            else{
                //if element in this sorted part
                if(nums[mid] <= target && target <= nums[e]){
                    s = mid+1;
                }
                //element not in this sorted part 
                else{
                    e = mid-1;
                }
            }
        }
        return -1;
    }